### ISSUE TITLE
Invalid memory access

### ISSUE THREAD

**Issue by ylansegal** [waterlink/spec2-mocks.cr#7]
Hey @waterlink, I am a crystal newbie: I believe I found an issue with spec2-mocks, although it's absolutely possible that I may be missing something obvious.

With the following code:

``` crystal
# test_spec.cr
require "spec2"
require "spec2-mocks"

module Foo
  VERSION = "0.1.0"
end

Spec2.describe Foo do
  it "has a version number" do
    expect(Foo::VERSION).not_to be_nil
  end
end
```

And my `shards.yml` looking like this:

``` yml
dependencies:
  spec2:
    github: waterlink/spec2.cr
    commit: eef5144d86f3aadf95811ffc0450be389db36fed
  spec2-mocks:
    github: waterlink/spec2-mocks.cr
    version: ~> 0.4
```

When I try to run, I get:

```
$ crystal test_spec.cr
Invalid memory access (signal 11) at address 0x0
[4572699642] *CallStack::print_backtrace:Int32 +42
[4559086963] __crystal_sigfault_handler +35
[140735744242986] _sigtramp +26
[4582303106] _ZN4llvm11PointerType3getEPNS_4TypeEj +26
[4578803049] _ZN4llvm17GetElementPtrInst16getGEPReturnTypeEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS4_EE +57
[4578320245] _ZN4llvm17GetElementPtrInst6CreateEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS4_EERKNS_5TwineEPNS_11InstructionE +117
[4577542886] _ZN4llvm9IRBuilderILb1ENS_14ConstantFolderENS_24IRBuilderDefaultInserterILb1EEEE17CreateInBoundsGEPEPNS_4TypeEPNS_5ValueENS_8ArrayRefIS8_EERKNS_5TwineE +138
[4581407291] LLVMBuildInBoundsGEP +54
[4567769571] *Crystal::CodeGenVisitor@Crystal::LLVMBuilderHelper#gep<LLVM::Value, Int32, Int32, String>:LLVM::Value +211
[4567963244] *Crystal::CodeGenVisitor#instance_var_ptr<Crystal::Type+, String, LLVM::Value>:LLVM::Value +1932
[4568052801] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +737
[4567942107] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1819
[4567941499] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1211
[4568007653] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1861
[4568031322] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +202
[4568128710] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +150
[4568078258] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +610
[4567941292] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1004
[4567941499] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1211
[4568007653] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1861
[4568031322] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +202
[4568128710] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +150
[4568078258] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +610
[4567941292] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1004
[4568113206] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +166
[4568078041] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +393
[4567941292] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1004
[4567941499] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1211
[4568350253] *Crystal::CodeGenVisitor#visit<Crystal::ExceptionHandler>:Bool +557
[4567943886] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +3598
[4568007653] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1861
[4568031322] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +202
[4568128710] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +150
[4568078258] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +610
[4568083679] *Crystal::CodeGenVisitor#codegen_dispatch<Crystal::Call, Array(Crystal::Def+)>:Bool +4207
[4568077720] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +72
[4567941292] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1004
[4567941499] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1211
[4568350253] *Crystal::CodeGenVisitor#visit<Crystal::ExceptionHandler>:Bool +557
[4567943886] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +3598
[4568265031] *Crystal::CodeGenVisitor#visit<Crystal::Yield>:Bool +2359
[4567941376] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1088
[4568265031] *Crystal::CodeGenVisitor#visit<Crystal::Yield>:Bool +2359
[4567941376] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1088
[4567941499] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1211
[4568240335] *Crystal::CodeGenVisitor#visit<Crystal::While>:Bool +607
[4567942522] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +2234
[4567941499] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1211
[4568121679] *Crystal::CodeGenVisitor#codegen_call_with_block<Crystal::Call, Crystal::Block, Crystal::Type+, Array(LLVM::Value)>:LLVM::Value +1215
[4568078282] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +634
[4567941292] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1004
[4568121679] *Crystal::CodeGenVisitor#codegen_call_with_block<Crystal::Call, Crystal::Block, Crystal::Type+, Array(LLVM::Value)>:LLVM::Value +1215
[4568078282] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +634
[4567941292] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1004
[4567941499] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1211
[4568350253] *Crystal::CodeGenVisitor#visit<Crystal::ExceptionHandler>:Bool +557
[4567943886] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +3598
[4568007653] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Module, Bool, Bool>:LLVM::Function +1861
[4568031322] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +202
[4568128710] *Crystal::CodeGenVisitor#codegen_call<Crystal::Call, Crystal::Def+, Crystal::Type+, Array(LLVM::Value)>:(Bool | LLVM::Value | Nil) +150
[4568078258] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +610
[4567941292] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1004
[4567941499] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1211
[4568007653] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, LLVM::Modul%
```

I am using crystal-lang 0.19.4 (which is why spec2 is pinned to a specific commit). 

If I remove the `require "spec2-mocks"` line, the spec runs as expected:

```
$ crystal test_spec.cr
.

Finished in 0.43 milliseconds
Examples: 1, failures: 0
```

Thank you for taking the time to write this library and supporting it. I hope I gave you all the information you need to debug this, but please do let me know if I can be of further help.

**Comment by waterlink** [comment]
My intuition tells me, that this problem will be resolved once you actually define some mocks. Probably some code in there does not expect that the collection of mocks is empty..

**Comment by alexanderadam** [comment]
I can confirm that you always get reproducible memory access errors when the mocks are removed.
This way users have no idea why their specs explode. 😉 

I can reproduce it with my [example project](https://github.com/alexanderadam/mocking_error/blob/master/spec/foo_spec.cr#L8-L10) if I remove the mock.

**Comment by waterlink** [comment]
Also, the error happens at compile time, not in runtime. Probably the most simple solution would be to add some unused mock in the library itself. What do you think?

**Comment by waterlink** [comment]
After fixing the bug #8, now it always fails with this error. Even the test suite of the library itself. Looking into that.

**Comment by alexanderadam** [comment]
An unused mock sounds good. I still have to think into these macro thingies…

**Comment by waterlink** [comment]
I have released `0.4.1` with the fix

**Comment by alexanderadam** [comment]
Here again: thank you very much 👍
